sunbeam-python/sunbeam/jobs/steps.py

```diff
@@ -0,0 +1,320 @@
+    _MODEL = None
+
+    def __init__(
+        self,
+        tfhelper: TerraformHelper,
+        jhelper: JujuHelper,
+        banner: str = "",
+        description: str = "",
+    ):
+        super().__init__(banner, description)
+        self.tfhelper = tfhelper
+        self.jhelper = jhelper
+        self.client = Client()
+
+    def _validate(self) -> bool:
+        if self._CONFIG and self._APPLICATION and self._MODEL:
+            return True
+
+        return False
+
+    def extra_tfvars(self) -> dict:
+        return {}
+
+    def get_application_timeout(self) -> int:
+        return 600
+
+    def is_skip(self, status: Optional[Status] = None) -> Result:
+        """Determines if the step should be skipped or not.
+
+        :return: ResultType.SKIPPED if the Step should be skipped,
+                ResultType.COMPLETED or ResultType.FAILED otherwise
+        """
+        if not self._validate():
+            return Result(ResultType.FAILED, "Required variables not set by the Step")
+
+        try:
+            run_sync(self.jhelper.get_application(self._APPLICATION, self._MODEL))
+        except ApplicationNotFoundException:
+            return Result(ResultType.COMPLETED)
+
+        return Result(ResultType.SKIPPED)
+
+    def run(self, status: Optional[Status] = None) -> Result:
+        """Apply terraform configuration to deploy sunbeam machine"""
+        machine_ids = []
+        try:
+            app = run_sync(self.jhelper.get_application(self._APPLICATION, self._MODEL))
+            machine_ids.extend(unit.machine.id for unit in app.units)
+        except ApplicationNotFoundException as e:
+            LOG.debug(str(e))
+
+        try:
+            tfvars = read_config(self.client, self._CONFIG)
+        except ConfigItemNotFoundException:
+            tfvars = {}
+
+        tfvars.update(self.extra_tfvars())
+        tfvars.update({"machine_ids": machine_ids})
+        update_config(self.client, self._CONFIG, tfvars)
+        self.tfhelper.write_tfvars(tfvars)
+
+        try:
+            self.tfhelper.apply()
+            print("seems like no exception raised")
+        except TerraformException as e:
+            return Result(ResultType.FAILED, str(e))
+
+        # Note(gboutry): application is in state unknown when it's deployed
+        # without units
+        try:
+            run_sync(
+                self.jhelper.wait_application_ready(
+                    self._APPLICATION,
+                    self._MODEL,
+                    accepted_status=["active", "unknown"],
+                    timeout=self.get_application_timeout(),
+                )
+            )
+        except TimeoutException as e:
+            LOG.warning(str(e))
+            return Result(ResultType.FAILED, str(e))
+
+        return Result(ResultType.COMPLETED)
+
+
+class AddMachineUnitStep(BaseStep):
+    """Base class to add unit of machine application"""
+
+    _CONFIG = None
+    _APPLICATION = None
+    _MODEL = None
```

Comment:
These should not be class attributes but part of the constructor and set at __init__ time. Simplifies logic  (`_validate` is no longer needed), and it becomes a linting error if they're missing.

This comment is applicable to multiple places

Reply:
Done